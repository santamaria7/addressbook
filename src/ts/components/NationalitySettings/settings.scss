.settings {
  form {
    width: 100%;
    max-width: 600px;
    margin: 0 auto;
  }
  label,
  select {
    display: block;
  }
  select {
    padding: 5px;
    min-width: 100px;
    margin: 10px 0;
  }
  button {
    padding: 5px;
    cursor: pointer;
  }
}
