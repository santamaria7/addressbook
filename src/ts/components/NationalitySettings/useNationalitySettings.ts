import { useDispatch, useSelector } from "react-redux";
import { useHistory } from "react-router-dom";
import { ChangeEvent, FormEvent, useState } from "react";
import { setNationalityAction } from "../../store/actions/setNationalityAction";
import { emptyUsersListAction } from "../../store/actions/emptyUsersListAction";

export const useNationalitySettings = () => {
  const dispatch = useDispatch();
  const { goBack } = useHistory();

  const n = useSelector<State>((state) => state.nationality) as string;

  const [value, setValue] = useState(n);
  const [showSuccess, setShowSuccess] = useState(false);

  const changeNatValue = (e: ChangeEvent<HTMLSelectElement>) => {
    setValue(e.target.value);
  };
  const submitForm = async (e: FormEvent) => {
    e.preventDefault();
    localStorage.setItem("nat", value);
    await dispatch(setNationalityAction(value));
    await dispatch(emptyUsersListAction());
    await setShowSuccess(true);
    goBack();
  };
  return {
    submitForm,
    value,
    changeNatValue,
    showSuccess,
  };
};
