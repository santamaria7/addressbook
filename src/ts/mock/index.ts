import configureMockStore from "redux-mock-store";
import createSagaMiddleware from "redux-saga";
import rootSaga from "../store/saga";
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
export function mockStore(){
    configure({ adapter: new Adapter() });
    const sagaMiddleware = createSagaMiddleware();
    const mockStoreAPI = configureMockStore([sagaMiddleware]);

    const store = mockStoreAPI({
        users: [],
        pagination: {
            limit: 1000,
            offset: 1,
        },
        search: {
            isFiltered: false,
        },
        loading: false,
        nationality: "",
    });
    sagaMiddleware.run(rootSaga);
    return store
}