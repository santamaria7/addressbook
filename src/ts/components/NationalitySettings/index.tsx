import * as React from "react";
import { ChangeEvent, FormEvent, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { setNationalityAction } from "../../store/actions/setNationalityAction";
import { useHistory } from "react-router-dom";
import { emptyUsersListAction } from "../../store/actions/emptyUsersListAction";
import "./settings.scss";

const NationalitySettings = () => {
  const dispatch = useDispatch();
  const { goBack } = useHistory();

  const n = useSelector<State>((state) => state.nationality) as string;

  const [value, setValue] = useState(n);
  const [showSuccess, setShowSuccess] = useState(false);

  const changeNatValue = (e: ChangeEvent<HTMLSelectElement>) => {
    setValue(e.target.value);
  };
  const submitForm = async (e: FormEvent) => {
    e.preventDefault();
    localStorage.setItem("nat", value);
    await dispatch(setNationalityAction(value));
    await dispatch(emptyUsersListAction());
    await setShowSuccess(true);
    goBack();
  };

  return (
    <div className="settings">
      <form onSubmit={submitForm}>
        <label>Set preferred nationality:</label>
        <select value={value} onChange={changeNatValue}>
          <option value="CH">CH</option>
          <option value="ES">ES</option>
          <option value="FR">FR</option>
          <option value="GB">GB</option>
        </select>
        <button type="submit">Save</button>
        {showSuccess && <p>Successfully Saved</p>}
      </form>
    </div>
  );
};

export default NationalitySettings;
