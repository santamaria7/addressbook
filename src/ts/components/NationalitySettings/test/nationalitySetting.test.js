/**  * @jest-environment jsdom-sixteen  */

import React from "react";
import renderer from "react-test-renderer";
import { render, fireEvent, screen } from "@testing-library/react";
import { mockStore } from "../../../../../mock";
import { Provider } from "react-redux";
import NationalitySettings from "../index";
import { waitFor } from "@testing-library/dom";

jest.mock("react-router-dom", () => {
  const originalModule = jest.requireActual("react-router-dom");
  return {
    ...originalModule,
    useHistory: () => ({
      goBack: jest.fn(),
    }),
  };
});

describe("NationalitySetting Component", function () {
  it("should render nationality settings successfully", function () {
    const component = renderer.create(
      <Provider store={mockStore()}>
        <NationalitySettings />
      </Provider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("should change the select value successfully", async function () {
    render(
      <Provider store={mockStore()}>
        <NationalitySettings />
      </Provider>
    );
    const element = screen.getByLabelText("Set preferred nationality:");
    fireEvent.change(element, { target: { value: "GB" } });
    await waitFor(() => {
      expect(element.value).toEqual("GB");
    });
  });
  it("should submit form successfully", async function () {
    render(
      <Provider store={mockStore()}>
        <NationalitySettings />
      </Provider>
    );
    const form = screen.getByRole("form");
    const value = screen.getByLabelText("Set preferred nationality:").value;
    fireEvent.submit(form);
    await waitFor(() => {
      //TODO don't use localStorage for test, always failing
      expect(localStorage.getItem("nat")).toEqual(value);
    });
  });
});
