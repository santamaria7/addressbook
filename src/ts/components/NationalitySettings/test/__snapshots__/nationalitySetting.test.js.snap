// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NationalitySetting Component should render nationality settings successfully 1`] = `
<div
  className="settings"
>
  <form
    onSubmit={[Function]}
    role="form"
  >
    <label
      htmlFor="select"
    >
      Set preferred nationality:
    </label>
    <select
      id="select"
      onChange={[Function]}
      value=""
    >
      <option
        value="CH"
      >
        CH
      </option>
      <option
        value="ES"
      >
        ES
      </option>
      <option
        value="FR"
      >
        FR
      </option>
      <option
        value="GB"
      >
        GB
      </option>
    </select>
    <button
      type="submit"
    >
      Save
    </button>
  </form>
</div>
`;
